### PR TITLE
Simplified Interfaces for Evaluation Functions and Added Previous States to SensorFusion

### LINKED ISSUE #1523: Refactor evaluation functions to take fewer arguments
### Description of the task
Many evaluation functions don't need World or Ball or Robot and can take a circular buffer of previous states or take a Timestamped_State

### Acceptance criteria

- [ ] as titled
- [ ] no regressions

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
Plan going forward: 
* SensorFusion will hold previous robot and ball states and will feed them into filter
* Will copy evaluation functions into sensorfusion as a filter before deleting them from evaluation and integrating them

### Resolved Issues
resolves #1523

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/sensor_fusion/sensor_fusion.cpp
Comment (akhilveeraghanta) [inline src/software/sensor_fusion/sensor_fusion.cpp]: We. should probably pull this into a const somewhere?

File: src/software/sensor_fusion/sensor_fusion.cpp
Comment (akhilveeraghanta) [inline src/software/sensor_fusion/sensor_fusion.cpp]: helper w/ team and robot state map as arguments will allow us to not repeat the same code, could be a private helper of sensor fusion?

File: src/software/ai/evaluation/robot.h
Comment (akhilveeraghanta) [inline src/software/ai/evaluation/robot.h]: should this live in `world/state.h`?

**Review by akhilveeraghanta — COMMENTED** [review]
⚽ 🌏

File: src/software/ai/evaluation/intercept.h
Comment (MathewMacDougall) [inline src/software/ai/evaluation/intercept.h:37]: nit: add blank line between throws and return

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Looking good overall :+1:

File: src/software/ai/evaluation/intercept.h
Comment (jonathanlew) [reply src/software/ai/evaluation/intercept.h:43]: hm... it's hard to know what the current time is (since Timestamps are in ssl-vision time), so Duration is much easier to use

File: src/software/sensor_fusion/sensor_fusion.cpp
Comment (jonathanlew) [reply src/software/sensor_fusion/sensor_fusion.cpp]: yes it should but that should be done at the same time as Robot and Ball, which will be refactored during this project, so I want to hold off on it

File: src/software/sensor_fusion/sensor_fusion.cpp
Comment (jonathanlew) [reply src/software/sensor_fusion/sensor_fusion.cpp]: good point

File: src/software/ai/evaluation/robot.h
Comment (jonathanlew) [reply src/software/ai/evaluation/robot.h]: it's only being used here in eval/robot.{h,cpp}, so I think it's best to limit the scope to here. Eventually, this will be moved to software/sensorfusion/[somewhere]

File: src/software/ai/evaluation/intercept.h
Comment (MathewMacDougall) [reply src/software/ai/evaluation/intercept.h:43]: Yea I think "Duration in future" is a good interface here.

File: src/software/ai/evaluation/intercept.h
Comment (jonathanlew) [reply src/software/ai/evaluation/intercept.h:37]: added

File: src/software/sensor_fusion/sensor_fusion.cpp
Comment (jonathanlew) [reply src/software/sensor_fusion/sensor_fusion.cpp]: cleaned up

**Comment by MathewMacDougall** [comment]
These merges really do be conflicting though :eyes:

**Comment by jonathanlew** [comment]
> These merges really do be conflicting though eyes

fixed dem merge conflicts

**Review by MathewMacDougall — DISMISSED** [review]

File: src/software/sensor_fusion/sensor_fusion.cpp
Comment (garethellis0) [inline src/software/sensor_fusion/sensor_fusion.cpp]: The `Team` data type is our abstraction around the filtered data that we pass to the AI. I'm a bit confused about why we're updating the `TeamHistoryMap` with a `Team` here? I would expect to use raw data to update the `TeamHistoryMap`, and then create the `Team`'s from that. 

I'm sure I'm missing something here, just trying to clarify.

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/software/ai/evaluation/robot.h
Comment (jonathanlew) [reply src/software/ai/evaluation/robot.h]: This will make it compatible with sensorfusion in the future because sensorfusion has circular buffers

File: src/software/ai/evaluation/intercept_test.cpp
Comment (jonathanlew) [reply src/software/ai/evaluation/intercept_test.cpp:253]: added test

File: src/software/ai/evaluation/robot.cpp
Comment (jonathanlew) [reply src/software/ai/evaluation/robot.cpp]: done

File: src/software/sensor_fusion/sensor_fusion.cpp
Comment (jonathanlew) [reply src/software/sensor_fusion/sensor_fusion.cpp]: I think that Team should have a map of id to robotstate, but we're not there yet. It's confusing to have it in this PR, so I'll remove it.

**Comment by jonathanlew** [comment]
Deleted more unused functions, please re-review

**Review by MathewMacDougall — DISMISSED** [review]

File: src/software/ai/evaluation/robot.h
Comment (garethellis0) [reply src/software/ai/evaluation/robot.h]: In that case, IMO this change really should come when we migrate these functions into sensorfusion. 

@akhilveeraghanta @MathewMacDougall for your thoughts

File: src/software/ai/evaluation/intercept.h
Comment (garethellis0) [inline src/software/ai/evaluation/intercept.h:43]: Looking at this again, I'm very confused why this function is not a part of the ball? It seems pretty logical that any ball model would be a part of the ball class?

File: src/software/ai/evaluation/robot.cpp
Comment (garethellis0) [reply src/software/ai/evaluation/robot.cpp]: Doesn't look like this changed?

**Review by garethellis0 — COMMENTED** [review]
.

File: src/software/ai/evaluation/intercept.h
Comment (EvanMorcom) [reply src/software/ai/evaluation/intercept.h:43]: IMO it would seem weird that the Ball is responsible for predicting its own future state.

As soon as we get a model that is more accurate than speed*time (like friction etc), I don't think it makes sense for the Ball class to have knowledge of that kind of information. @MathewMacDougall

**Review by EvanMorcom — DISMISSED** [review]
LGTM 🐶

File: src/software/ai/evaluation/robot.cpp
Comment (jonathanlew) [reply src/software/ai/evaluation/robot.cpp]: my bad changed now

File: src/software/ai/evaluation/intercept.h
Comment (jonathanlew) [reply src/software/ai/evaluation/intercept.h:43]: I agree with Evan's that estimate future time is a utility function (think geom types) that doesn't belong in `Ball`. The longterm goal of this rearch is that Ball will become effectively TimestampedBallState and the current functionality will live in evaluation functions or in SensorFusion filters. This PR is already 1500 lines and I haven't changed anything functional. I thought it would be a good place to stop before I move more things around. With this PR a lot of things in Ball and Robot are dismantled. `Ball::esimatePositionAtFutureTime` was only being used in intercept so I pulled it out here. I can inline it if you feel that's an improvement.

File: src/software/ai/evaluation/robot.h
Comment (MathewMacDougall) [reply src/software/ai/evaluation/robot.h]: No strong opinion. This change still matches the theme of this PR by simplifying the interface to just the buffers of data it needs, rather than full `Ball` and `Robot` types

File: src/software/ai/evaluation/intercept.h
Comment (MathewMacDougall) [reply src/software/ai/evaluation/intercept.h:43]: Hmmm I see the arguments for both sides here (function in/outside of the class). Trying to separate my bias of being "familiar" with it as part of the class.

If for example we stored a ball model in some Dynamic Params in the future, those could easily be passed to the Ball class or a free function, so I don't see a huge advantage one way or the either there.

Would be good to hear from @LiCody @jared-paul or @kelvinkoon as to what they find more intuitive too.

Comment (jonathanlew) [reply src/software/ai/evaluation/intercept.h:43]: I think what we agreed on with the rearch was that `World` would contain a lightweight data type to represent the ball, e.g. `TimestampedBallState` and it would not contain history or any additional functionality (same for TimestampedRobotState). Ideas like possession and BallModel should be handled separately. So BallModel can be its own class if that makes sense when we implement it, but it won't be the ball class that gets passed from SensorFusion to AI.

**Review by EvanMorcom — APPROVED** [review]

**Review by MathewMacDougall — APPROVED** [review]
LGTM pending disussion on ball model and prediction

File: src/software/ai/evaluation/intercept.h
Comment (garethellis0) [reply src/software/ai/evaluation/intercept.h:43]: I might have misspoke during our discussion, but my intent was that `World` holds an abstraction around the ball, including being able to get it's estimated future state. I was advocating for the removal of history, which seemed like an implementation detail.

File: src/software/ai/evaluation/robot.h
Comment (garethellis0) [reply src/software/ai/evaluation/robot.h]: So I still disagree with this change, but if we're going to move this to sensorfusion in the immediate future then I won't hold this PR up.

File: src/software/ai/evaluation/intercept.h
Comment (EvanMorcom) [reply src/software/ai/evaluation/intercept.h:43]: I feel like adding state prediction to World would require a name change to the class. 'World' implies a world state at a moment vs fluidity of time IMO.

Comment (garethellis0) [reply src/software/ai/evaluation/intercept.h:43]: I mean, we need to somehow convey this model to the AI without the raw buffers being exposed.

Comment (jonathanlew) [reply src/software/ai/evaluation/intercept.h:43]: Why does ball prediction need inputs from a state history buffer? I thought that ball prediction can be done statically from a physics model of the ball and the ball state (velocity + position). If so, then this ball model can live entirely in AI

Comment (EvanMorcom) [reply src/software/ai/evaluation/intercept.h:43]: There are some ball prediction algorithms that would require history. Based solely on friction we could determine future states from the current state, but any curvature (due to rotating dribblers/kickers) you would need previous states to judge the curve

Comment (jonathanlew) [reply src/software/ai/evaluation/intercept.h:43]: That's a good point since we'll have rotating kickers soon. I think @MathewMacDougall point that we want BallState to be as simple as possible for so we don't have to mock things that we don't need is relevant here. SensorFusion should define a separate BallModel class to pass into AI rather than piggyback on ball state. I can work on that structure in #1522

Comment (garethellis0) [reply src/software/ai/evaluation/intercept.h:43]: Bit confused here, are we suggesting that `World` would have `BallState` and `BallModel` as independent objects? 

(_might be faster to have this discussion on slack_)

Comment (jonathanlew) [reply src/software/ai/evaluation/intercept.h:43]: An important point here is to keep `World` to only representing a snapshot of the world, since it is used in `Simulation`. After offline discussion, there will be a need to combine `world_state`, `possession`, `ball_model`, and other modules into a `SensorFusion` output for `AI` to use. `estimateBallPositionAtFutureTime` will be moved once when we work on the `BallModel`

Comment (garethellis0) [reply src/software/ai/evaluation/intercept.h:43]: I can't comment on whether or not this is needed for Sim (though I suspect it's not), but I'm really not sure where this is going in terms of the output of sensor fusion. I certainly don't think it makes any sense to break up the `Ball` class, perhaps it might make sense to have `Ball` contain a `BallState` and a `BallModel`?

Comment (MathewMacDougall) [reply src/software/ai/evaluation/intercept.h:43]: > An important point here is to keep World to only representing a snapshot of the world, since it is used in Simulation

World has no bearing on simulation. Simulation just outputs raw SSL data, it has no knowledge of the World.

I think `Ball` containing a ballstate and ball model makes sense, because it puts it at a similar level of abstraction to the new (soon TM) Robot class that is basically state with "extra metadata"

Comment (jonathanlew) [reply src/software/ai/evaluation/intercept.h:43]: OK that makes sense, I'll make a BallModel and put it in Ball in the next PR with #1522

**Review by akhilveeraghanta — APPROVED** [review]
LGTM 🤖

**Review by garethellis0 — APPROVED** [review]

**Pull request merged**
